File: backend-tanaw/app/core/password_policy.py
import json
import unicodedata
from pathlib import Path
from typing import cast


def _load_password_policy_data() -> dict[str, object]:
    raw_data = json.loads(
        Path(__file__).with_name("common_passwords.json").read_text(encoding="utf-8")
    )
    if not isinstance(raw_data, dict):
        raise RuntimeError("The password policy data must be an object.")
    return cast(dict[str, object], raw_data)


def _required_int(data: dict[str, object], key: str) -> int:
    value = data.get(key)
    if not isinstance(value, int) or isinstance(value, bool) or value <= 0:
        raise RuntimeError(f"The password policy value {key!r} is invalid.")
    return value


def _required_string(data: dict[str, object], key: str) -> str:
    value = data.get(key)
    if not isinstance(value, str) or not value:
        raise RuntimeError(f"The password policy value {key!r} is invalid.")
    return value


def _required_strings(data: dict[str, object], key: str) -> tuple[str, ...]:
    values = data.get(key)
    if not isinstance(values, list) or not all(isinstance(value, str) for value in values):
        raise RuntimeError(f"The password policy values {key!r} are invalid.")
    return tuple(cast(list[str], values))


_PASSWORD_POLICY_DATA = _load_password_policy_data()
PASSWORD_MIN_LENGTH = _required_int(_PASSWORD_POLICY_DATA, "minimumLength")
PASSWORD_MAX_LENGTH = _required_int(_PASSWORD_POLICY_DATA, "maximumLength")
PASSWORD_TOO_SHORT_MESSAGE = _required_string(_PASSWORD_POLICY_DATA, "tooShortMessage")
PASSWORD_TOO_LONG_MESSAGE = _required_string(_PASSWORD_POLICY_DATA, "tooLongMessage")
PASSWORD_COMMON_MESSAGE = _required_string(_PASSWORD_POLICY_DATA, "commonMessage")
_COMMON_PASSWORD_STEMS = _required_strings(_PASSWORD_POLICY_DATA, "stems")
_COMMON_PASSWORD_EXACT = _required_strings(_PASSWORD_POLICY_DATA, "exact")
_COMMON_PASSWORD_SUFFIXES = _required_strings(_PASSWORD_POLICY_DATA, "suffixes")
_COMMON_PASSWORD_SEPARATORS = _required_strings(_PASSWORD_POLICY_DATA, "separators")


def normalize_password(value: str) -> str:
    return unicodedata.normalize("NFC", value)


def validate_password_policy(value: str) -> str:
    normalized = normalize_password(value)
    length = len(normalized)
    if length < PASSWORD_MIN_LENGTH:
        raise ValueError(PASSWORD_TOO_SHORT_MESSAGE)
    if length > PASSWORD_MAX_LENGTH:
        raise ValueError(PASSWORD_TOO_LONG_MESSAGE)
    if is_common_password(normalized):
        raise ValueError(PASSWORD_COMMON_MESSAGE)
    return normalized


def is_common_password(value: str) -> bool:
    normalized = normalize_password(value).casefold()
    return not normalized.strip() or normalized in COMMON_PASSWORD_BLOCKLIST


def _build_common_password_blocklist() -> frozenset[str]:
    candidates = {normalize_password(value).casefold() for value in _COMMON_PASSWORD_EXACT}
    for stem_value in _COMMON_PASSWORD_STEMS:
        stem = normalize_password(stem_value).casefold()
        for repeat_count in range(2, 5):
            for separator in _COMMON_PASSWORD_SEPARATORS:
                candidate = separator.join(stem for _ in range(repeat_count))
                if PASSWORD_MIN_LENGTH <= len(candidate) <= PASSWORD_MAX_LENGTH:
                    candidates.add(candidate)
        for suffix in _COMMON_PASSWORD_SUFFIXES:
            candidate = f"{stem}{suffix}"
            if PASSWORD_MIN_LENGTH <= len(candidate) <= PASSWORD_MAX_LENGTH:
                candidates.add(candidate)
    return frozenset(candidates)


COMMON_PASSWORD_BLOCKLIST = _build_common_password_blocklist()

File: desktop-tanaw/src/utils/password-policy.ts
import passwordPolicyData from "../data/commonPasswords.json";

export const PASSWORD_MIN_LENGTH = passwordPolicyData.minimumLength;
export const PASSWORD_MAX_LENGTH = passwordPolicyData.maximumLength;
export const PASSWORD_INPUT_MAX_CODE_UNITS = PASSWORD_MAX_LENGTH * 2;
export const PASSWORD_TOO_SHORT_MESSAGE = passwordPolicyData.tooShortMessage;
export const PASSWORD_TOO_LONG_MESSAGE = passwordPolicyData.tooLongMessage;
export const PASSWORD_COMMON_MESSAGE = passwordPolicyData.commonMessage;

const COMMON_PASSWORD_BLOCKLIST = buildCommonPasswordBlocklist();

export type PasswordRequirementStatus = {
  characterCount: number;
  hasValue: boolean;
  isLengthValid: boolean;
  isNotCommon: boolean;
};

export function normalizePassword(value: string) {
  return value.normalize("NFC");
}

export function validatePasswordPolicy(value: string) {
  const status = getPasswordRequirementStatus(value);
  if (status.characterCount < PASSWORD_MIN_LENGTH) return PASSWORD_TOO_SHORT_MESSAGE;
  if (status.characterCount > PASSWORD_MAX_LENGTH) return PASSWORD_TOO_LONG_MESSAGE;
  if (!status.isNotCommon) return PASSWORD_COMMON_MESSAGE;
  return "";
}

export function getPasswordRequirementStatus(value: string): PasswordRequirementStatus {
  const normalized = normalizePassword(value);
  const characterCount = Array.from(normalized).length;
  const hasValue = characterCount > 0;
  return {
    characterCount,
    hasValue,
    isLengthValid: characterCount >= PASSWORD_MIN_LENGTH && characterCount <= PASSWORD_MAX_LENGTH,
    isNotCommon: hasValue && Boolean(normalized.trim()) && !COMMON_PASSWORD_BLOCKLIST.has(normalized.toLowerCase()),
  };
}

function buildCommonPasswordBlocklist() {
  const candidates = new Set(passwordPolicyData.exact.map((value) => normalizePassword(value).toLowerCase()));
  for (const stemValue of passwordPolicyData.stems) {
    const stem = normalizePassword(stemValue).toLowerCase();
    for (let repeatCount = 2; repeatCount < 5; repeatCount += 1) {
      for (const separator of passwordPolicyData.separators) {
        const candidate = Array.from({ length: repeatCount }, () => stem).join(separator);
        const length = Array.from(candidate).length;
        if (length >= PASSWORD_MIN_LENGTH && length <= PASSWORD_MAX_LENGTH) candidates.add(candidate);
      }
    }
    for (const suffix of passwordPolicyData.suffixes) {
      const candidate = `${stem}${suffix}`;
      const length = Array.from(candidate).length;
      if (length >= PASSWORD_MIN_LENGTH && length <= PASSWORD_MAX_LENGTH) candidates.add(candidate);
    }
  }
  return candidates;
}
